runtimebp/cpu.go

```diff
@@ -12,88 +12,40 @@ import (
 // NumCPU returns the number of CPUs assigned to this running container.
 //
 // This is the container aware version of runtime.NumCPU.
-// It reads from the cgroup sysfs values.
+// It reads from the cgroup cpu.cfs_quota_us and cpu.cfs_period_us values
+// to determine the hard CPU limit of the container.
 //
 // If the current process is not running inside a container,
-// or for whatever reason we failed to read the cgroup sysfs values,
+// there is no cpu.cfs_quota_us, or if it failes to read the cgroup values,
 // it will fallback to runtime.NumCPU() instead.
-//
-// When fallback happens, it also prints the reason to stderr.
-func NumCPU() (n float64) {
+func NumCPU() float64 {
 	const (
 		quotaPath  = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us"
 		periodPath = "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
 	)
 
-	var err error
-	defer func() {
-		if err != nil || n <= 0 {
-			// Fallback and log to stderr.
-			fmt.Fprintf(
-				os.Stderr,
-				"NumCPU: falling back to use shares: %v, %v\n",
-				n,
-				err,
-			)
-			n = numCPUSharesFallback()
-		}
-	}()
+	// Default to the standard runtime package value.
+	defaultCPUs := float64(runtime.NumCPU())
 
 	// Big enough buffer to read the number in the file wholly into memory.
 	buf := make([]byte, 1024)
-	var quota, period float64
 
-	quota, err = readNumberFromFile(quotaPath, buf)
+	quota, err := readNumberFromFile(quotaPath, buf)
 	if err != nil {
-		return
+		return defaultCPUs
 	}
 
-	period, err = readNumberFromFile(periodPath, buf)
-	if err != nil {
-		return
+	// CFS quota returns -1 if there is no limit, return the default.
+	if quota < 0 {
+		return defaultCPUs
 	}
 
-	return quota / period
-}
-
-// On some situations the quota file will be -1,
-// and we should use this one instead.
-//
-// Those situations include:
-//
-// - Very old version of docker
-//
-// - In k8s only request is set for cpu, not limit
-func numCPUSharesFallback() (n float64) {
-	const (
-		sharesPath  = "/sys/fs/cgroup/cpu/cpu.shares"
-		denominator = 1024
-	)
-
-	var err error
-	defer func() {
-		if err != nil || n <= 0 {
-			// Fallback and log to stderr.
-			fmt.Fprintf(
-				os.Stderr,
-				"NumCPU: falling back to use runtime.NumCPU(): %v, %v\n",
-				n,
-				err,
-			)
-			n = float64(runtime.NumCPU())
-		}
-	}()
-
-	// Big enough buffer to read the number in the file wholly into memory.
-	buf := make([]byte, 1024)
-	var shares float64
-
-	shares, err = readNumberFromFile(sharesPath, buf)
+	period, err := readNumberFromFile(periodPath, buf)
 	if err != nil {
-		return
+		return defaultCPUs
 	}
 
-	return float64(shares) / denominator
+	return quota / period
```

Comment:
🔕 If this is the code path we hit for .5 cpu limit, then it'll round down to 0, so this would be `(quota+quota-1) / period` if we want to round up.

Reply:
this function returns a float number (so it returns 0.5 instead of rounding), [`runtimebp.GOMAXPROCS`](https://pkg.go.dev/github.com/reddit/baseplate.go/runtimebp#GOMAXPROCS) is the one actually using it and do the rounding, and it rounds up.